### ISSUE TITLE
classes.jsa and classes_nocoops.jsa now excluded from MacOS JDKs/JREs?

### ISSUE THREAD

**Issue by chadlwilson** [adoptium/adoptium-support#937]
## Question

I noticed that the JREs for `17.0.9` now exclude `classes*.jsa` where they were included in `17.0.8`.

While smaller JREs distributions are always great, my understanding was that class data sharing was still an important part of improving JRE startup time. Can someone point me to the reading and impact of this compared to previous releases? I infer perhaps these were not actually being used/effective earlier, or perhaps removed as part of the reproducible build effort.

## Context

Java version:

```
openjdk version "17.0.9" 2023-10-17
OpenJDK Runtime Environment Temurin-17.0.9+9 (build 17.0.9+9)
OpenJDK 64-Bit Server VM Temurin-17.0.9+9 (build 17.0.9+9, mixed mode)
```

Your operating system and platform:

MacOS Sonoma 14.1 on aarch64 (but relevant for other platforms)

**Comment by karianna** [comment]
I've asked our build team to investigate

**Comment by chadlwilson** [comment]
After looking a little bit further (but not exhaustively across platforms and archs) it sort of looks like

- Mac aarch64 builds: `classes*.jsa` has been been excluded for quite a while, at least versions back as far as `17.0.4`
- Mac x64 builds: `classes*.jsa` started getting excluded from `17.0.9`+

So I guess they are at least consistent now. But not sure of impact.

**Comment by chadlwilson** [comment]
Based on a recent [Mac aarch64 build log](https://ci.adoptium.net/job/build-scripts/job/jobs/job/jdk17u/job/jdk17u-mac-aarch64-temurin/lastSuccessfulBuild/console) perhaps might be something to do with cross-compilation if relying on default archives being created.

Perhaps aarch64 was always cross-compiled (so no defualt Archive), and from `17.0.9`+ x64 was also considered cross-compiled due to some build env changes moving it to Apple Silicon machines?

I note work to possibly move to arm64 machines and cross-compile x64 (which _might_ explain why the archives disappeared for x64 builds) in places like https://github.com/adoptium/infrastructure/issues/2536 but can't really make sense of all the WIP as an outsider :-)

**Comment by chadlwilson** [comment]
I'm pretty sure this is still a problem.

**Comment by giger85** [comment]
Is there any update?

**Comment by sxa** [comment]
> I note work to possibly move to arm64 machines and cross-compile x64 (which might explain why the archives disappeared for x64 builds) in places like https://github.com/adoptium/infrastructure/issues/2536 but can't really make sense of all the WIP as an outsider :-)

Hi - to confirm what you said yes we are cross compiling the x64 ones on an aarch64 mac and have been for about the last year. Having said that, it looks like we haven't had the shared class cache included in aarch64 mac for any of the jdk17 releases as 17+35 also did not include it. As you say, 17.0.8.1+1 was the last JDK17 mac release to include it, which I believe is consistent with when we switched over to the x64 cross-compiled builds.

Also noting that this is not specific to the JREs - the JDK tarballs are also missing them (filenames here were shortened for my own convenience but 11 and 21 are the latest versions. The 17 tarballs have the exact version number on them):
```
sxa@orangepi5plus:/dev/shm$ for A in *gz; do echo ===== $A; tar tfz $A | grep classes; done
===== 11amac.tgz
===== 11xmac.tgz
===== 170amac.tgz
===== 170xmac.tgz
jdk-17+35/Contents/Home/lib/server/classes_nocoops.jsa
jdk-17+35/Contents/Home/lib/server/classes.jsa
===== 1781amac.tgz
===== 1781xmac.tgz
jdk-17.0.8.1+1/Contents/Home/lib/server/classes_nocoops.jsa
jdk-17.0.8.1+1/Contents/Home/lib/server/classes.jsa
===== 179amac.tgz
===== 179xmac.tgz
===== 21amac.tgz
===== 21xmac.tgz
```

**Comment by chadlwilson** [comment]
Thanks! I'm a little surprised that no-one else on Mac x64 noticed this changing, but maybe that just reflects that most people had been on Apple Silicon/aarch64 for ages, which has never had the archives included.

While I do not know of the desired Temurin compliance with things like [JEP 345](https://openjdk.org/jeps/341), and each vendor can presumably make their own decisions, it is worth noting that this is a little inconsistent here and I imagine if these could be included MacOS performance would be improved relative to linux aarchs (which seem to be there for Alpine and regular Linux across x64 and aarch64 at least) and also compared to other distributions.

**Comment by sxa** [comment]
> Thanks! I'm a little surprised that no-one else on Mac x64 noticed this changing

Yeah I'm quite surprised too :-)

> it is worth noting that this is a little inconsistent here

Absolutely - I don't believe this is intentional and I'll be discussing it with the team tomorrow, although we have some people on vacation at this time of year. Hopefully we can get something in place before the next release though. Also it's an easy thing for us to put into some of our checks so I'll make sure that's done too if we get this resolved.

> seems some possible quirk of the build env that is causing the archives to neither aarch64 nor x64 and both to be considered as cross-compiled for the purposes of this.

Yes I agree - we should be consistent and I suspect we'll need to understand why they're not currently being produced so thanks for pushing this one. As you suggest, the cross-compilation /shouldn't/ be relevant here since it's failing on the natively built aarch64 one too. I'm not entirely sure what the implications might be of creating an x64 CDS archive under Rosetta though ... It might not generate something optimal.

The [latest build log for aarch64](https://ci.adoptium.net/job/build-scripts/job/jobs/job/jdk21u/job/jdk21u-mac-aarch64-temurin/148/consoleFull) initially looks ok:
```
20:49:08  checking if platform is supported by CDS... yes
20:49:08  checking if JVM feature 'cds' is available... yes
```
But later on in the build it seems to skip the generation because it thinks it's cross-compiling, so hopefully we can sort it out!
```
20:49:08  checking if CDS archive is available... no (not possible with cross compilation)
20:49:08  checking if a default CDS archive should be generated... disabled, from default 'auto'
20:49:08  checking if CDS archive is available... yes
20:49:08  checking if compatible cds region alignment enabled... disabled, default
```
The x64 one has the same:
```
03:07:22  checking if the CDS classlist generation should be enabled... enabled, from default 'auto'
03:07:22  checking if any translations should be excluded... no
03:07:22  checking if static man pages should be copied... enabled, default
03:07:22  checking if CDS archive is available... no (not possible with cross compilation)
03:07:22  checking if a default CDS archive should be generated... disabled, from default 'auto'
03:07:22  checking if CDS archive is available... yes
```

**Comment by sxa** [comment]
@chadlwilson FYI in the meantime you should be able to populate the shared class files on your local machine with:
`java -Xshare:dump`
And if you need the nocoops version:
`java -Xshare:dump -XX:-UseCompressedOops`

_(As an aside, doing this is generally a nicer option anyway since the cache files are always somewhat dependent on the machine you're running on, so doing it on your local machine might give you something that's more suitable and efficient)_

**Comment by sxa** [comment]
Remove the `PMC-agenda` label today as this was discussed at the PMC meeting and @gdams took an action to replicate and investigate what's needed to resolve this at least for aarch64 on his system.

**Comment by chadlwilson** [comment]
Thanks @sxa - am aware of the workaround from reading the JEPs and earlier knowledge.

https://github.com/gocd/gocd redistributes/bundles Temurin JREs across platforms for users. MacOS is not so common a production deployment target so not critical for me to bother adding special logic and aarch64 builders for this platform.

Locally l probably switch around [mise](https://mise.jdx.dev/)-managed JREs and JDKs too often locally to worry about doing this consistently.

I can't even 100% remember what I was doing when I noticed this - probably just diffing some GoCD binary distributions to compare between releases at some point and wondering why our distribution got smaller between releases with different JRE bundled versions.

And then down the CDS and JEP rabbithole through (morbid?) curiosity :-)

**Comment by sxa** [comment]
> I can't even 100% remember what I was doing when I noticed this

:-) But good that you did notice it since it gives us a chance to fix it and make sure we can trap it if the cache files disappear in the future.

**Comment by gdams** [comment]
Noting that running configure on my local macOS M1 laptop returns `checking if CDS archive is available... yes`

**Comment by sxa** [comment]
> Noting that running configure on my local macOS M1 laptop returns `checking if CDS archive is available... yes`

Yep that line is also present in the log snippets in [my earlier comment](https://github.com/adoptium/adoptium-support/issues/937#issuecomment-2301698941) from the CI

**Comment by chadlwilson** [comment]
Thanks folks!

I have no objection, but ..should we take it that generating CDS archives for MacOS x64 is a "won't fix" given the available build hardware (Apple Silicon) and platform priority?

**Comment by sxa** [comment]
> Thanks folks!
> 
> I have no objection, but ..should we take it that generating CDS archives for MacOS x64 is a "won't fix" given the available build hardware (Apple Silicon) and platform priority?

At the moment it's more that it was an easy fix to resolve it on aarch64 so we got that applied. The question world be how to generate something meaningful in the cross-compiled case. Is there any value in a van created in the Risers environment for example, or would we need to run the generation on x64?

This issue was closed because the related PR auto closed it. I'm going to reopen it until we have closure one way or another on x64 too, but it will require someone to do some investigation I think.

**Comment by headius** [comment]
Is there a plan to fix this? I'm running into this on JRuby where we are trying to smartly add in AppCDS flags to automatically generate the archive if it doesn't exist. Unfortunately now if we're running on Temurin on macos on Apple Silicon, we have to detect that we're running with a JDK that doesn't even have its own dump. This is really inconvenient.

I also tried to force a dump using `-Xshare:dump` and it doesn't seem to have helped. It's going to be a pain to ship this feature for users when we know there's going to be some that install Temurin without the base CDS archive available.

**Comment by karianna** [comment]
Hmm, I can't actually see where we set or unset this in our build scripts, only one small reference where we add an extra parameter for older versions of 17 on Linux.  Over to @gdams to see if he knows more.

**Comment by chadlwilson** [comment]
@headius I didn't specifically verify this for recent releases but in theory the archives should now be there for aarch64/Apple Silicon (but still missing for x64).

Did you try with the Jan 2025 patch release distributions?

**Comment by headius** [comment]
@chadlwilson Fair point, I'm sure I have not updated these in many months. I'll try the most recent builds and get back to you.

It never occurred to me that this feature would not work if the host JDK was not also using CDS so we may have to build in that detection anyway.

**Comment by chadlwilson** [comment]
FWIW, I can confirm that both archives are there in Temurin 21.0.5+ mac aarch64 archives (and some others I happen to have locally).

```
$ find  ~/.local/share/mise/installs/java -name "*.jsa"
/Users/chad/.local/share/mise/installs/java/temurin-17.0.13+11/lib/server/classes_nocoops.jsa
/Users/chad/.local/share/mise/installs/java/temurin-17.0.13+11/lib/server/classes.jsa
/Users/chad/.local/share/mise/installs/java/temurin-17.0.14+7/lib/server/classes_nocoops.jsa
/Users/chad/.local/share/mise/installs/java/temurin-17.0.14+7/lib/server/classes.jsa
/Users/chad/.local/share/mise/installs/java/temurin-21.0.5+11.0.LTS/lib/server/classes_nocoops.jsa
/Users/chad/.local/share/mise/installs/java/temurin-21.0.5+11.0.LTS/lib/server/classes.jsa
/Users/chad/.local/share/mise/installs/java/temurin-21.0.6+7.0.LTS/lib/server/classes_nocoops.jsa
/Users/chad/.local/share/mise/installs/java/temurin-21.0.6+7.0.LTS/lib/server/classes.jsa
```

**Comment by neilccbrown** [comment]
Just a quick observation that although the .jsa files are now back in the MacOS aarch build, they are missing from the x64 build.  I can understand that the aarch build is more important going forward, but just thought I'd point this out for a full understanding.

**Comment by chadlwilson** [comment]
While it's possibly misleading as an issue now, the issue was indeed left open due to the missing x64 versions: https://github.com/adoptium/adoptium-support/issues/937#issuecomment-2312015504

Unless there is hardware available to the project to support a x64 build without a cross-compile from Apple Silicon I imagine this will be stuck for a while. :-/

**Comment by headius** [comment]
Perhaps an Open Source account from MacStadium might work? They are mostly Apple Silicon but they do still host some i7 instances. We used an OSS MacStadium account for testing JRuby on Apple Silicon before it was widely available.

https://www.macstadium.com/company/opensource

Eventually, though, this issue will be moot, as there will be no supported x64 Mac instances available anywhere.
